### PR TITLE
[Feature] New parser and analyzer for `grant`/`revoke` privileges

### PR DESCRIPTION
**Pull request by padmejin**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

* Use new parser & analyzer for the following SQL:
```
GRANT SELECT,INSERT ON dbxx.tablexx TO userxx;
GRANT SELECT_PRIV ON *.* TO ROLE rolexx;
GRANT USAGE ON RESOURCE * TO userxxx;
GRANT USAGE ON RESOURCE 'spark0' TO 'userxxx';
```
* `grant impersonate` will be processed as common grant statement.

This PR  also introduces the following side effects.
* To grant or revoke privileges to a role that doesn't exist is not allowed anymore, nor will the corresponding role be created automatically. We abandoned this usage because we believe it is unreasonable.
* The action name postfix '_PRIV' is not required anymore.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 249 / 274 (90.88%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|[160]|
|:large_blue_circle:|com/starrocks/execution/GrantExecutor.java|0|1|00.00%|[14]|
|:large_blue_circle:|com/starrocks/execution/RevokeExecutor.java|0|1|00.00%|[14]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|19|25|76.00%|[588, 596, 601, 608, 616, 622]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzer.java|51|62|82.26%|[124, 129, 138, 154, 164, 169, 170, 171, 172, 184, 201]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|71|75|94.67%|[3166, 3204, 3208, 4325]|
|:large_blue_circle:|com/starrocks/sql/ast/BaseGrantRevokePrivilegeStmt.java|41|42|97.62%|[109]|
|:large_blue_circle:|com/starrocks/mysql/privilege/ResourcePrivEntry.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/GrantPrivilegeStmt.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/AccessPrivilege.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Auth.java|16|16|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/ImpersonateUserPrivEntry.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/RevokePrivilegeStmt.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/GlobalPrivEntry.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/TablePrivEntry.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AST2SQL.java|24|24|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/DbPrivEntry.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/execution/DataDefinitionExecutorFactory.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 252 / 269 (93.68%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|[165]|
|:large_blue_circle:|com/starrocks/execution/GrantExecutor.java|0|1|00.00%|[14]|
|:large_blue_circle:|com/starrocks/execution/RevokeExecutor.java|0|1|00.00%|[14]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|19|25|76.00%|[590, 598, 603, 610, 618, 624]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzer.java|56|62|90.32%|[126, 131, 140, 156, 166, 186]|
|:large_blue_circle:|com/starrocks/sql/ast/BaseGrantRevokePrivilegeStmt.java|41|42|97.62%|[109]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|69|70|98.57%|[3378]|
|:large_blue_circle:|com/starrocks/mysql/privilege/ResourcePrivEntry.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/GrantPrivilegeStmt.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/AccessPrivilege.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Auth.java|16|16|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/ImpersonateUserPrivEntry.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/RevokePrivilegeStmt.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/GlobalPrivEntry.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/TablePrivEntry.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AST2SQL.java|24|24|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/DbPrivEntry.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/execution/DataDefinitionExecutorFactory.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 252 / 269 (93.68%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|[167]|
|:large_blue_circle:|com/starrocks/execution/GrantExecutor.java|0|1|00.00%|[14]|
|:large_blue_circle:|com/starrocks/execution/RevokeExecutor.java|0|1|00.00%|[14]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|19|25|76.00%|[600, 608, 613, 620, 628, 634]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzer.java|56|62|90.32%|[127, 132, 141, 157, 167, 187]|
|:large_blue_circle:|com/starrocks/sql/ast/BaseGrantRevokePrivilegeStmt.java|41|42|97.62%|[109]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|69|70|98.57%|[3400]|
|:large_blue_circle:|com/starrocks/mysql/privilege/ResourcePrivEntry.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/GrantPrivilegeStmt.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/AccessPrivilege.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Auth.java|16|16|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/ImpersonateUserPrivEntry.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/RevokePrivilegeStmt.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/GlobalPrivEntry.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/TablePrivEntry.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AST2SQL.java|24|24|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/DbPrivEntry.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/execution/DataDefinitionExecutorFactory.java|2|2|100.00%|[]|

**Review by chaoyli — DISMISSED** [review]

**Comment by padmejin** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 251 / 268 (93.66%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|[167]|
|:large_blue_circle:|com/starrocks/execution/GrantExecutor.java|0|1|00.00%|[14]|
|:large_blue_circle:|com/starrocks/execution/RevokeExecutor.java|0|1|00.00%|[14]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|19|25|76.00%|[600, 608, 613, 620, 628, 634]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzer.java|55|61|90.16%|[127, 132, 141, 157, 167, 187]|
|:large_blue_circle:|com/starrocks/sql/ast/BaseGrantRevokePrivilegeStmt.java|41|42|97.62%|[109]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|69|70|98.57%|[3400]|
|:large_blue_circle:|com/starrocks/mysql/privilege/ResourcePrivEntry.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/GrantPrivilegeStmt.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/AccessPrivilege.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Auth.java|16|16|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/ImpersonateUserPrivEntry.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/RevokePrivilegeStmt.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/GlobalPrivEntry.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/TablePrivEntry.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AST2SQL.java|24|24|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/DbPrivEntry.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/execution/DataDefinitionExecutorFactory.java|2|2|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java [hunk 1]
```diff
@@ -3274,29 +3275,143 @@ public ParseNode visitRevokeRoleStatement(StarRocksParser.RevokeRoleStatementCon
...
+        }
+        return stmt;
+    }
+
+    @Override
+    public ParseNode visitPrivilegeAction(StarRocksParser.PrivilegeActionContext context) {
+        if (context.privilegeActionReserved() != null) {
+            return new StringLiteral(context.privilegeActionReserved().getText());
+        } else {
+            return new StringLiteral(context.identifier().getText());
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java:3410]: visit(context.identifier())

**Review by HangyuanLiu — APPROVED** [review]

**Review by gengjun-git — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by padmejin** [comment]
run starrocks_fe_unittest

**Comment by padmejin** [comment]
run starrocks_fe_unittest

**Comment by padmejin** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 250 / 267 (93.63%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|[168]|
|:large_blue_circle:|com/starrocks/execution/GrantExecutor.java|0|1|00.00%|[14]|
|:large_blue_circle:|com/starrocks/execution/RevokeExecutor.java|0|1|00.00%|[14]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|19|25|76.00%|[608, 616, 621, 628, 636, 642]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzer.java|55|61|90.16%|[127, 132, 141, 157, 167, 187]|
|:large_blue_circle:|com/starrocks/sql/ast/BaseGrantRevokePrivilegeStmt.java|41|42|97.62%|[109]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|68|69|98.55%|[3324]|
|:large_blue_circle:|com/starrocks/mysql/privilege/ResourcePrivEntry.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/GrantPrivilegeStmt.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/AccessPrivilege.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Auth.java|16|16|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/ImpersonateUserPrivEntry.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/RevokePrivilegeStmt.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/GlobalPrivEntry.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/TablePrivEntry.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AST2SQL.java|24|24|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/DbPrivEntry.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/execution/DataDefinitionExecutorFactory.java|2|2|100.00%|[]|

**Pull request merged**